drivers/scheduler/anthos/anthos.go

```diff
@@ -529,6 +543,126 @@ func (anth *anthos) unsetUserNameAndKey() error {
 	return nil
 }
 
+// checkUserClusterNodesUpgradeTime measure the time taken by each node and report error
+func (anth *anthos) checkUserClusterNodesUpgradeTime() error {
+	log.Info("Validating user cluster nodes upgrade time")
+	userCluster, err := anth.getUserClusterName()
+	if err != nil {
+		return err
+	}
+	initNodeUpgradeTime, err := anth.getStartTimeForNodePoolUpgrade(userCluster)
+	if err != nil {
+		return err
+	}
+	log.Debugf("User cluster node pool upgrade started at: [%v]", initNodeUpgradeTime.Format(time.UnixDate))
+	poolMap, err := getNodesSortByAge()
+	if err != nil {
+		return err
+	}
+	for _, sortedNodes := range poolMap {
+		startTime := initNodeUpgradeTime
+		for _, node := range sortedNodes {
+			diff := node.CreationTimestamp.Sub(startTime)
+			if diff > errorTimeDuration {
+				return fmt.Errorf("[%s] node upgrade took: [%v] minutes which is longer than the expected timeout value: [%v]",
+					node.Name, diff, errorTimeDuration)
+			}
+			log.Infof("[%s] node took: [%v] time to upgrade the node", node.Name, diff)
+			startTime = node.CreationTimestamp.Time
+		}
+	}
+	return nil
+}
+
+// getUserClusterName return Anthos user cluster name
+func (anth *anthos) getUserClusterName() (string, error) {
+	log.Info("Retrieving user cluster name")
+	var userCluster string
+	// Listing user cluster to get user cluster name
+	cmd := fmt.Sprintf("%s --kubeconfig %s clusters", listUserClustersCmd, adminKubeconfPath)
+	out, err := anth.execOnAdminWSNode(cmd)
+	if err != nil {
+		return "", fmt.Errorf("listing user clusters is failing: [%s]. Err: (%v)", out, err)
+	}
+	userClusters := strings.Split(out, "\n")
+	for _, cluster := range userClusters {
+		if strings.Contains(cluster, "NAME") {
```

Comment:
You don't need this once you use the new command with grep.

Reply:
Incorporated !!